### PR TITLE
Unittests for InternalNamesTest.h

### PR DESCRIPTION
**Pull request by Sameeranjoshi**
I have 2 queries:
`line 103`: Should be 4(DEFAULT KIND) ?
`line 138`: Why moD1 is not lower case?

I have added a TODO as some functionality of `deConstruct` is WIP.
`ninja check-flang-unit` should fail as of this PR, will get a fix once the above query is resolved.

### CODE REVIEW

File: flang/unittests/Optimizer/InternalNamesTest.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+TEST(InternalNamesTest, doIntrinsicTypeDescriptorTest) {
+  NameUniquer obj;
+  std::string actual =
+      obj.doIntrinsicTypeDescriptor({}, {}, obj.IntrinsicType::REAL, {42});
+  std::string expectedMangledName = "_QCrealK42";
+  ASSERT_EQ(actual, expectedMangledName);
+
+  actual = obj.doIntrinsicTypeDescriptor({}, {}, obj.IntrinsicType::REAL, {});
+  expectedMangledName = "_QCrealK0"; // [FIXME]: Should be 4(DEFAULT KIND) ?
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/InternalNamesTest.cpp:103]: The KindType is just an integer for this interface, so when you use `{}` as a ctor (e.g., `std::int64_t{}`) you will get the value zero.  Kinds are expected to come from the Fortran front-end. We wouldn't expect C++ to know about them. (The braces in the other calls like line 99 aren't needed, btw.)

File: flang/unittests/Optimizer/InternalNamesTest.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  ASSERT_EQ(actual, expectedMangledName);
+}
+
+TEST(InternalNamesTest, doTypeDescriptorTest) {
+  NameUniquer obj;
+  std::string actual = obj.doTypeDescriptor(
+      {StringRef("moD1")}, {StringRef("foo")}, "MyTYPE", {2, 8});
+  std::string expectedMangledName =
+      "_QMmod1FfooCTmytypeK2K8"; // [FIXME]: Why moD1 is not lower case?,
+                                 // expected: "_QMmod1FfooCTmytypeK2K8"
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/InternalNamesTest.cpp:140]: That's a bug. Looks like your test is doing well already.

File: flang/unittests/Optimizer/CMakeLists.txt [hunk 3]
```diff
@@ -1,30 +1,13 @@
...
-  MLIRStandardToLLVM
-  MLIRTransforms
-  MLIRAffineToStandard
-  MLIRAnalysis
-  MLIRSCFToStandard
-  MLIREDSC
-  MLIRParser
-  MLIRStandardToLLVM
-  MLIRSupport
-  MLIRVectorToLLVM
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/CMakeLists.txt:21]: Where these libraries needed for `Basic.cpp`? Will they be needed for other unit tests?

File: flang/unittests/Optimizer/InternalNamesTest.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+  std::string expectedMangledName = "_QQ@MAIN";
+  ASSERT_EQ(actual, expectedMangledName);
+
+  std::string actual1 = obj.doGenerated("@_ZNSt8ios_base4InitC1Ev");
+  std::string expectedMangledName1 = "_QQ@_ZNSt8ios_base4InitC1Ev";
+  ASSERT_EQ(actual1, expectedMangledName1);
+
+  std::string actual2 = obj.doGenerated("_QQ@MAIN");
+  std::string expectedMangledName2 = "_QQ_QQ@MAIN";
+  ASSERT_EQ(actual2, expectedMangledName2);
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/InternalNamesTest.cpp:71]: +1  Good to test that it all works with symbols too.

File: flang/unittests/Optimizer/CMakeLists.txt (hunk 3 above)
Comment (Sameeranjoshi) [reply flang/unittests/Optimizer/CMakeLists.txt:21]: No idea about dependency for other tests.
I figured out later that `${dialect_libs}` does the work.

**Comment by schweitzpgi** [comment]
I had to revert this because of compilation errors (14).
```
flang/unittests/Optimizer/InternalNamesTest.cpp:98:64: error: 
      'fir::NameUniquer::IntrinsicType::REAL' is not a member of class
      'fir::NameUniquer'
      obj.doIntrinsicTypeDescriptor({}, {}, obj.IntrinsicType::REAL, 42);
                                                ~~~~~~~~~~~~~~~^
```

**Comment by Sameeranjoshi** [comment]
> I had to revert this because of compilation errors (14).
> 
> ```
> flang/unittests/Optimizer/InternalNamesTest.cpp:98:64: error: 
>       'fir::NameUniquer::IntrinsicType::REAL' is not a member of class
>       'fir::NameUniquer'
>       obj.doIntrinsicTypeDescriptor({}, {}, obj.IntrinsicType::REAL, 42);
>                                                 ~~~~~~~~~~~~~~~^
> ```

I wasn't able to reproduce above error.
But I saw some FIR related build errors, when I checkout to commit before merging #98 
Then I pulled fir-dev upstream changes.
I re-based my commits(3 commits) on top of fir-dev, I can see all builds and tests passing.

**Pull request merged**